### PR TITLE
Swap only optimal amount even when active ID moves

### PR DESCRIPTION
**Pull request by jihoonsong**
Handling edge case that active ID moves during swap on preparing params. The goal of this algorithm is to find the optimal amount `S`.

- Calculate all the necessary parameters including how much of which token to swap. (Let's call the amount to swap here as `S1`)
- If the swap will move active ID, then swap only amount, `S2`,  as much as need to move active ID. And restart this process with the new active ID. (Note that we must swap at least `S2`, otherwise we can't provide liquidity.)
- If no swap is required or the swap doesn't move active ID, we have all the right parameters including the optimal amounts of X and Y. It's done.

This leads us to optimal swap because `S` is the summation of `S2`s, which is equal to or smaller than `S1`.

### CODE REVIEW

**Comment by jihoonsong** [comment]
If you set block height `24080000` and run tests, you can see an active ID goes up during deposit. To see an active ID goes down, you can use the same block height and run `should succeed when adjust to consecutive bins below active bin` test under `testTraderJoeAdjustBins`.

**Comment by ferencdg** [comment]
So in the worst case scenario, we might swap 24 or 25 times right? Do we know the gas cost for that? We need to know because the stable coin pair price might above/under the range in which we provide liquidity, and we don't want traderJoe strategy to prevent deposits into the calm portfolio. (because of the block gas limit)

**Comment by ferencdg** [comment]
I like that you tried to provide the algorithm that allows us to invest the entire amount of the user in one go. However whether we can go live with that or not  depends on the gas cost of the worst-case scenario, so we need to check that.

If that worst-case scenario doesn't work, we can just implement a simple one.
1. calculate the distribution 
2. do a swap even if it changes the active bins it is fine
3. re-calculate the distribution, however at this point we assume we will do no more swaps in the future
4. provide liquidity (we might have quite a bit of leftover assets)
5. swap the non-deposit currency back to deposit currency
6. let the off-chain component reinvest it

**Comment by ferencdg** [comment]
this uses less gas in the worst case, but do unnecessary swapping and delays the full investment until the offchain component is kicked in

**Comment by jihoonsong** [comment]
The active ID has different affect on deposit algorithm whether it is on target bins, any bin below target bins, and any bin above them. Let's say that there are $n$ target bins. Then, there are $n + 2$ possibilities that active ID resides on, $t_0$, $t_1$, ..., $t_n$, $t_{n+1}$ where $t_0$ is any bin below target bins, $t_1$, ..., $t_n$ are the target bins, and $t_{n+1}$ is any bin above target bins.

As long as an active ID goes up by 1 bin, it is good to add liquidity as is. We just end up with the fact that we swapped more than necessary. The amount of excess swap is $ExcessSwapAmount = TotalSwapAmount - MinAmountToMoveActiveID$, which is $\ge 0$.

What if it moves by more than 1 bin? We can't add liquidity because we don't have liquidity for bins between previous active ID and current one (exclusive). Let's call the sum of allocations for those bins $S$. We need to swap back, but swapping back $S$ can also make active ID to move and we can't predict how many bins it will pass this time. If it goes more than 1 bin, we need to swap again. In order to swap back without any further trouble, we need to find the amount that can be allocated to bins between min target bin and equal to or below than the active one after swap by iterating each bin down from current active one. (or vice versa.) Sounds complex but can be implemented with one for loop.

Before implement this one, let's see how often this case will occur as of current condition.

i) active ID is below 8388607
  Deposit amount needs to be around 0.25M plus USDC.e reserve of bins below 8388607.
ii) active ID is 8388607
  Deposit amount needs to be around 0.5M plus USDC.e reserve of 8388607.
iii) active ID is 8388608
  Deposit amount needs to be around 0.1M plus USDC.e reserve of 8388608.
iv) active ID is 8388609 or above it
 No swap occurs.

Currently, the pool's TVL is 1.5M. A $0.1M deposit to TraderJoe, or $1M deposit to CALM equivalently, under certain situation is unlikely. So I will go with a decision to make it revert when an active ID moves more than 1 bin.

**Comment by jihoonsong** [comment]
The worst case is iterate max target bin's ID - active ID times where active ID is below target bins (or active ID - min target bin's ID in the other way.) (Also FYI, bins are finite.)

Each of 10 bins around and including the target bins have a larger TVL than that of the CALM portfolio. Extending it, each of 20 bins have a larger TVL than what is probably the max deposit amount in all of CALM histories. Also, TraderJoe receives only 10% of the total CALM deposit.

So considering above all with current gas usage, active ID's deviation in the stablecoin pool, and whatever else, it's not likely that this will go over the block gas limit.

File: contracts/strategies/traderjoe/TraderJoeInvestmentLib.sol [hunk 1]
```diff
@@ -224,6 +224,30 @@ library TraderJoeInvestmentLib {
...
+        amountIn = SwapServiceLib.swapTokensForExactTokens(
+            strategyStorage.swapService,
+            amountOut,
+            type(uint256).max,
+            path,
+            binSteps
+        );
+    }
+
     function __prepareParams(uint256 amountXIn, uint256 amountYIn)
```
[the remark this answers is not in the document]
Comment (jihoonsong) [reply contracts/strategies/traderjoe/TraderJoeInvestmentLib.sol:251]: Compound, Aave, OpenZepplin, Python, and `__gap` in this repo.

**Comment by jihoonsong** [comment]
So it is good enough solution, especially taking our product's situation into account. Nonetheless, I updated it with better one.

The active ID has different affect on deposit algorithm whether it is on target bins, any bin below target bins, and any bin above them. Let's say that there are $n$ target bins. Then, there are $n + 2$ possibilities that active ID resides on, $t_0$, $t_1$, ..., $t_n$, $t_{n+1}$ where $t_0$ is any bin below target bins, $t_1$, ..., $t_n$ are the target bins, and $t_{n+1}$ is any bin above target bins.

As long as an active ID goes up by 1 bin, it is good to add liquidity as is. We just end up with the fact that we swapped more than necessary. The amount of excess swap is $ExcessSwapAmount = TotalSwapAmount - MinAmountToMoveActiveID$, which is $\ge 0$.

What if it moves by more than 1 bin? We can't add liquidity because we don't have liquidity for bins between previous active ID and current one (exclusive). Let's call the sum of allocations for those bins $S$. We need to swap back, but swapping back $S$ can also make active ID to move and we can't predict how many bins it will pass this time. If it goes more than 1 bin, we need to swap again. In order to swap back without any further trouble, we need to find the amount that can be allocated to bins between min target bin and equal to or below than the active one after swap by iterating each bin down from current active one. (or vice versa.) Sounds complex but can be implemented with one for loop.

Before implement this one, let's see how often this case will occur as of current condition.

i) active ID is below 8388607
  Deposit amount needs to be around 0.25M plus USDC.e reserve of bins below 8388607.
ii) active ID is 8388607
  Deposit amount needs to be around 0.5M plus USDC.e reserve of 8388607.
iii) active ID is 8388608
  Deposit amount needs to be around 0.1M plus USDC.e reserve of 8388608.
iv) active ID is 8388609 or above it
 No swap occurs.

Currently, the pool's TVL is 1.5M. A $0.1M deposit to TraderJoe, or $1M deposit to CALM equivalently, under certain situation is unlikely. So I will go with a decision to make it revert when an active ID moves more than 1 bin.

**Pull request closed without merging**